src/Azure.Functions.Cli/Helpers/ZipHelper.cs

```diff
@@ -87,21 +94,117 @@ public static bool GoZipExists(out string fileLocation)
-            return fileStream;
+
+            // In order to properly mount and/or unzip this in Azure, we need to create the zip as if it were
+            // Unix so that the correct file permissions set above are applied. To do this, we walk backwards
+            // through the stream and update the "created by" field to 3, which indicates it was created by Unix.
+            if (OperatingSystem.IsWindows())
+            {
+                memStream.Seek(0, SeekOrigin.End);
+
+                // Update the file header in the zip file for every file to indicate that it was created by Unix
+                while (SeekBackwardsToSignature(memStream, centralDirectorySignature))
+                {
+                    // The field we need to set is 5 bytes from the beginning of the signature. Set it,
+                    // then move back to the previous location so we can continue.
+                    memStream.Seek(5, SeekOrigin.Current);
+                    memStream.WriteByte(CreatedByUnix);
+                    memStream.Seek(-6, SeekOrigin.Current);
+                }
+            }
+
+            memStream.Seek(0, SeekOrigin.Begin);
+            return memStream;
+        }
+
+        // Assumes all bytes of signatureToFind are non zero.
+        // If the signature is found then returns true and positions stream at first byte of signature
+        // If the signature is not found, returns false
+        private static bool SeekBackwardsToSignature(Stream stream, uint signatureToFind)
+        {
+            int bufferPointer = 0;
+            uint currentSignature = 0;
+            byte[] buffer = new byte[32];
+            bool signatureFound = false;
+
+            while (!signatureFound)
+            {
+                bufferPointer = SeekBackwardsAndRead(stream, buffer);
+
+                if (bufferPointer == -1)
+                {
+                    break;
+                }
+
+                while (bufferPointer >= 0 && !signatureFound)
+                {
+                    currentSignature = (currentSignature << 8) | ((uint)buffer[bufferPointer]);
+                    if (currentSignature == signatureToFind)
+                    {
+                        signatureFound = true;
+                    }
+                    else
+                    {
+                        bufferPointer--;
+                    }
+                }
+            }
+
+            if (!signatureFound)
+            {
+                return false;
+            }
+            else
+            {
+                stream.Seek(bufferPointer, SeekOrigin.Current);
+                return true;
+            }
+        }
+
+        // Returns true if we are out of bytes
+        private static int SeekBackwardsAndRead(Stream stream, byte[] buffer)
+        {
+            if (stream.Position >= buffer.Length)
+            {
+                stream.Seek(-buffer.Length, SeekOrigin.Current);
+                stream.Read(buffer.AsSpan());
```

Comment:
Let's switch this to `ReadExactly` to make the intent clear.

Reply:
Any real benefit of using a span here?

Reply:
We chatted about it, but yeah, `Read` and `ReadExactly` need a `Span`